File: python/Reverse Linked List.py
__author__ = 'BK'
'''
Reverse a singly linked list.

click to show more hints.

Subscribe to see which companies asked this question
'''
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head :return head
        p,q=head,head.next
        p.next=None

        while q:
            temp,q.next=q.next,p
            p,q=q,temp
        return p

def tral(node):
    if node:
        print node.val
        tral(node.next)

solu = Solution()
a=ListNode(1)
b=ListNode(2)
c=ListNode(3)
d=ListNode(4)
e=ListNode(5)
a.next=b
b.next=c
c.next=d
d.next=e
tral(a)

tral(solu.reverseList(a))

File: leetcode/Reverse Linked List.cpp
/*
Reverse a linked list from position m to n. Do it in-place and in one-pass.

For example:
Given 1->2->3->4->5->NULL, m = 2 and n = 4,

return 1->4->3->2->5->NULL.

Note:
Given m, n satisfy the following condition:
1 ¡Ü m ¡Ü n ¡Ü length of list.

Subscribe to see which companies asked this question
*/ 
#include<bits/stdc++.h>
using namespace std;
 struct ListNode {
     int val;
     ListNode *next;
     ListNode(int x) : val(x), next(NULL) {}
 };

ListNode* helper(ListNode* curr,ListNode* prev,ListNode* head)
{
	if (!curr->next)
	{
		curr->next=prev;
		return curr;
	}
	ListNode* next = curr->next;
	curr->next = prev;
	helper(next,curr,head);
}

ListNode* helper2(ListNode* head)
{
	if (!head)return head;
	ListNode* p=head;
	ListNode* q=head->next;
	ListNode* temp=NULL;
	p->next=NULL;
	
	while(q)
	{
		temp=q->next;
		q->next=p;
		
		p=q;
		q=temp;
	}
	return p;
}

void print(ListNode* head)
{
	if (!head)
	{
		cout<<endl;
		return;
	}
	cout<<head->val<<" ";
	print(head->next);
}

int main()
{
	ListNode* a=new ListNode(1);
	a->next=new ListNode(2);
	a->next->next=new ListNode(3);
	a->next->next->next=new ListNode(4);
	a->next->next->next->next=new ListNode(5);
	print (a);
	ListNode* head = a;
	print(helper2(head));
	//print(a);
}
